odoo/fields.py

```diff
@@ -3047,6 +3052,629 @@ def _record_ids_per_res_model(self, records):
+                if not current_model or res_model != current_model:
+                    # the model on the child does not match the model on the parent
+                    # maybe we have changed the model on the parent, we should reset
+                    # the many2one value
+                    property_value = False
+                elif check_existence and res_id:
+                    # many2one might have been deleted
+                    property_value = env[res_model].browse(res_id).exists().id
+                else:
+                    property_value = res_id
+
+            elif property_type == 'many2many' and property_value:
+                # [[1337, 1338, 1339], 'res.partner']
+                assert check_list_types(property_value, [(list, tuple), str]) \
+                    and is_list_of(property_value[0], int), \
+                    f'Wrong many2many value: {property_value!r}.'
+
+                res_ids, res_model = property_value
+
+                if len(res_ids) != len(set(res_ids)):
+                    # remove duplicated value and preserve order
+                    res_ids = list(dict.fromkeys(res_ids))
+
+                if not current_model or res_model != current_model:
+                    # the model on the child does not match the model on the parent
+                    # maybe we have changed the model on the parent, we should reset
+                    # the many2many value
+                    property_value = False
+                elif check_existence and res_ids:
+                    # many2many might have been deleted
+                    property_value = env[res_model].browse(res_ids).exists().ids
+                else:
+                    property_value = res_ids
+
+            property_definition['value'] = property_value
+
+    @classmethod
+    def _list_to_dict(cls, values_list):
+        """Convert a list of properties with definition into a dict {name: value}.
+
+        To not repeat data in database, we only store the value of each property on
+        the child. The properties definition is stored on the parent.
+
+        E.G.
+            Input list:
+            [{
+                'name': 'discussion_color_code',
+                'string': 'Color Code',
+                'type': 'char',
+                'default': 'blue',
+                'value': 'red',
+            }, {
+                'name': 'moderator_partner_id',
+                'string': 'Partner',
+                'type': 'many2one',
+                'comodel': 'test_new_api.partner',
+                'value': [1337, 'Bob'],
+            }]
+
+            Output dict:
+            {
+                'discussion_color_code': 'red',
+                'moderator_partner_id': (1337, "res.partner"),
+            }
+
+        :param values_list: List of properties definition and value
+        :return: Generate a dict {name: value} from this definitions / values list
+        """
+        assert is_list_of(values_list, dict)
```

Comment:
Don't rely on `assert` for dynamic checks on production.  They will not be executed.